<a href="https://colab.research.google.com/github/Wasabibish/CLA_HW/blob/main/Hyper_params_selection_Homework.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from sklearn.datasets import load_digits
from sklearn import svm
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import classification_report

import autosklearn.classification as automl


**Dataset to be used is the Digits dataset from sklearn**

In [2]:
digits = load_digits()
digits.data.shape
X = digits.data
y = digits.target

In [3]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)     

**First use the support vector machine classifier along with Grid search to get the best SVC classifier possible for this task**

In [4]:
%%time
parameters = {'kernel':('linear', 'rbf'), 'C':[1, 10]}
svc = svm.SVC()
model = GridSearchCV(svc, parameters)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

CPU times: user 688 ms, sys: 6.91 ms, total: 695 ms
Wall time: 699 ms


In [5]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        55
           1       0.98      1.00      0.99        55
           2       0.98      1.00      0.99        52
           3       1.00      0.96      0.98        56
           4       1.00      1.00      1.00        64
           5       0.99      0.99      0.99        73
           6       1.00      1.00      1.00        57
           7       0.98      0.98      0.98        62
           8       0.98      0.98      0.98        52
           9       0.97      0.97      0.97        68

    accuracy                           0.99       594
   macro avg       0.99      0.99      0.99       594
weighted avg       0.99      0.99      0.99       594



In [6]:
model.best_params_

{'C': 10, 'kernel': 'rbf'}

**Next you will be using the auto sklearn to find the best fitting classifier for this dataset**

In [7]:
%%time
cls = automl.AutoSklearnClassifier(time_left_for_this_task=60)
cls.fit(X_train, y_train)
predictions = cls.predict(X_test)

CPU times: user 2.25 s, sys: 721 ms, total: 2.97 s
Wall time: 56.3 s


In [8]:
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        55
           1       0.98      1.00      0.99        55
           2       1.00      1.00      1.00        52
           3       0.98      0.98      0.98        56
           4       1.00      1.00      1.00        64
           5       0.97      1.00      0.99        73
           6       1.00      0.98      0.99        57
           7       0.98      0.98      0.98        62
           8       0.98      0.98      0.98        52
           9       0.98      0.96      0.97        68

    accuracy                           0.99       594
   macro avg       0.99      0.99      0.99       594
weighted avg       0.99      0.99      0.99       594



In [14]:
print(cls.sprint_statistics())

auto-sklearn results:
  Dataset name: e944b9e4-a920-11ec-8173-0242ac1c0002
  Metric: accuracy
  Best validation score: 0.989924
  Number of target algorithm runs: 14
  Number of successful target algorithm runs: 10
  Number of crashed target algorithm runs: 0
  Number of target algorithms that exceeded the time limit: 3
  Number of target algorithms that exceeded the memory limit: 1



**Compare the results in terms of (execution time, model complexity, and model performance)**



---



> **Execution Time**


*   SVC Classifier: 699 ms
*   Auto Classifier: 56.3 s

We notite thatAuto classifier takes more time, because it chooses an algorithm that fits our data between algorithms


---


> **Model Complexity**

*   SVC Classifier:  
*   Auto Classifier:

I didn't get what is complexity, is it like (time, space complexity) or the complexity as if our model is simple or cemplex

---


> **Model Performance**

*   SVC Classifier: 0.99 accuracy
*   Auto Classifier: 0.99 accuracy

Both algorithms performed well, with an accuracy equal to 0.99

